### PR TITLE
Merge in Jason's dev changes

### PR DESCRIPTION
**Pull request by blisc**
- Update Codec Autocast
- Add in Xuesong's clone fix
- Add in latest codec changes
- Updates epsilon logic 
- Adds ability to pass in nemo file for magpie in infer script

### CODE REVIEW

File: nemo/collections/tts/modules/audio_codec_modules.py [hunk 1]
```diff
@@ -407,40 +417,41 @@ def forward(self, x, l2_norm=False):
         Shapes:
             - x: :math:`(N, 1, T_{in})` or :math:`(N, D_{spec}, T_{in})`
         """
-        x.squeeze_(1)
-        # if you torch spec compute it otherwise use the mel spec computed by the AP
-        if self.use_torch_spec:
-            x = self.torch_spec(x)
+        with default_precision(torch.float32):
```
Comment (paarthneekhara) [inline nemo/collections/tts/modules/audio_codec_modules.py:420]: Is this different from using: torch.autocast(device_type=str(audio.device), dtype=torch.float32) ?

Comment (Edresson) [reply nemo/collections/tts/modules/audio_codec_modules.py:420]: Yes, it is. We adopted the default precision using the context manager approach, as that’s how PyTorch Lightning handles bf16 precision.

[half.py (lines 54–67)](https://github.com/Lightning-AI/pytorch-lightning/blob/a0134d238a45b3094c07bd0fa717a117858410dd/src/lightning/pytorch/plugins/precision/half.py#L54-L67)

[precision.py (lines 168–190)](https://github.com/Lightning-AI/pytorch-lightning/blob/a0134d238a45b3094c07bd0fa717a117858410dd/src/lightning/pytorch/plugins/precision/precision.py#L168-L190)

I learned this from Piotr.

File: nemo/collections/tts/models/magpietts.py [hunk 2]
```diff
@@ -284,39 +284,39 @@ def audio_to_codes(self, audio, audio_len, audio_type='target'):
...
+            # Make a copy to avoid modifying the original tensor if it's used elsewhere
+            codes_copy = codes.clone()
+            # Replace eos and bos tokens with padding in the copied tensor
+            codes_copy[codes == self.audio_bos_id] = 0  # zero is the padding token
+            codes_copy[codes == self.audio_eos_id] = 0
+            # Pass the modified integer token IDs
+            audio, audio_len = self._codec_model.decode(tokens=codes_copy, tokens_len=codes_len)
+            # audio: (B, T)
+            # audio_len: (B,)
+            return audio, audio_len
```
Comment (XuesongYang) [inline nemo/collections/tts/models/magpietts.py]: 1. replace `device_type=str(codes.device)` with `device_type=codes.device.type`. the former might pass the full device string (e.g., `'cuda:0'`) instead of just the type (`'cuda'`). `torch.autocast` primarily expects the type (`'cuda'`, `'cpu'`, etc.).
2. recommend to disable autocast functionality within context block since the default precision for most PyTorch operations is already `torch.float32`. And we **only** need to disable autocast for teh internal op of the codec decoding.

Sharing my suggestion:

```suggestion
        with torch.no_grad():
            # Make a copy to avoid modifying the original tensor if it's used elsewhere
            codes_copy = codes.clone()
            # Replace eos and bos tokens with padding in the copied tensor
            codes_copy[codes == self.audio_bos_id] = 0  # zero is the padding token
            codes_copy[codes == self.audio_eos_id] = 0
            
            # Force fp32 for the internal operations of the codec decoding
            with torch.autocast(device_type=codes.device.type, enabled=False):
                # Pass the modified integer token IDs. 
                # audio: (B, T), audio_len: (B,)
                audio, audio_len = self._codec_model.decode(tokens=codes_copy, tokens_len=codes_len)
                
            return audio, audio_len
```

File: nemo/collections/tts/models/magpietts.py [hunk 3]
```diff
@@ -284,39 +284,39 @@ def audio_to_codes(self, audio, audio_len, audio_type='target'):
...
+                (codes.size(0), codes.size(1), 1), 0, dtype=codes.dtype, device=codes.device
+            )  # 0 is the padding token in the audio codebook
+            codes = torch.cat([bos_tensor, codes, pad_tensor], dim=-1)
+            # codes: (B, C, T')
+            # codes_len: (B,)
+            for idx in range(codes.size(0)):
+                codes[idx, :, codes_len[idx] + 1] = audio_eos_id
+            codes_len = codes_len + 2
+
+            return codes.long(), codes_len.long()
```
Comment (XuesongYang) [inline nemo/collections/tts/models/magpietts.py]: This may be not a good place to cast to fp32 if `audio` has already been cast to bf16 before feeding into the function `def audio_to_codes()`. casting a tensor from fp32 to bf16 and then back to fp32 will generally result in a loss of precision. So we should disable the cast much earlier before calling `self.audio_to_codes()`. Wrapp disabling autocast at some places where `audio` was loaded.

File: scripts/magpietts/infer_and_evaluate.py [hunk 4]
```diff
@@ -54,33 +73,36 @@ def run_inference(
...
+        raise ValueError("Need a checkpoint")
 
-    # Load weights from checkpoint file
-    print("Loading weights from checkpoint")
-    ckpt = torch.load(checkpoint_file, weights_only=False)
-    model.load_state_dict(ckpt['state_dict'])
     print("Loaded weights.")
     model.cuda()
     model.eval()
     # import ipdb; ipdb.set_trace()
```
Comment (XuesongYang) [inline scripts/magpietts/infer_and_evaluate.py]: remove.

File: scripts/magpietts/infer_and_evaluate.py [hunk 5]
```diff
@@ -54,33 +73,36 @@ def run_inference(
...
+        checkpoint_name = nemo_file.split(".nemo")[0]
+    else:
+        raise ValueError("Need a checkpoint")
 
-    # Load weights from checkpoint file
-    print("Loading weights from checkpoint")
-    ckpt = torch.load(checkpoint_file, weights_only=False)
-    model.load_state_dict(ckpt['state_dict'])
     print("Loaded weights.")
     model.cuda()
```
Comment (XuesongYang) [inline scripts/magpietts/infer_and_evaluate.py:104]: is it necessary? if calling `ModelPT.restore_from()` without specifying `map_location=cpu`, the model should be cuda ready.

File: nemo/collections/tts/losses/audio_codec_loss.py [hunk 6]
```diff
@@ -230,8 +230,8 @@ def output_types(self):
     @typecheck()
     def forward(self, audio_real, audio_gen, audio_len):
         spec_len = (audio_len // self.hop_length) + 1
-        spec_real = self._compute_spectrogram(audio=audio_real, spec_len=spec_len)
-        spec_gen = self._compute_spectrogram(audio=audio_gen, spec_len=spec_len)
+        spec_real = self._compute_spectrogram(audio=audio_real.float(), spec_len=spec_len).to(audio_gen.dtype)
+        spec_gen = self._compute_spectrogram(audio=audio_gen.float(), spec_len=spec_len).to(audio_gen.dtype)
```
Comment (XuesongYang) [inline nemo/collections/tts/losses/audio_codec_loss.py:234]: similar reason as mentioned when using bf16-mixed, autocast to bf16 should be disabled so keeping `audio_real` or `audio_real` with fp32. Otherwise, cast to bf16 and back to fp32 results in loss of precision. We should disable autocast to bf16 much earlier before calling forward.

File: nemo/collections/tts/models/audio_codec.py [hunk 7]
```diff
@@ -594,8 +609,8 @@ def on_train_epoch_end(self):
     def validation_step(self, batch, batch_idx):
         audio, audio_len, audio_gen, _ = self._process_batch(batch)
 
-        loss_mel_l1, loss_mel_l2 = self.mel_loss_fn(audio_real=audio, audio_gen=audio_gen, audio_len=audio_len)
-        loss_stft = self.stft_loss_fn(audio_real=audio, audio_gen=audio_gen, audio_len=audio_len)
+        loss_mel_l1, loss_mel_l2 = self.mel_loss_fn(audio_real=audio.float(), audio_gen=audio_gen.float(), audio_len=audio_len)
+        loss_stft = self.stft_loss_fn(audio_real=audio.float(), audio_gen=audio_gen.float(), audio_len=audio_len)
```
Comment (XuesongYang) [inline nemo/collections/tts/models/audio_codec.py:613]: same comments for potential loss of precision.

File: nemo/collections/tts/modules/audio_codec_modules.py [hunk 8]
```diff
@@ -407,40 +417,41 @@ def forward(self, x, l2_norm=False):
...
+            # if you torch spec compute it otherwise use the mel spec computed by the AP
+            if self.use_torch_spec:
+                x = self.torch_spec(x)
 
-        if self.log_input:
-            x = (x + 1e-6).log()
-        x = self.instancenorm(x).unsqueeze(1)
+            if self.log_input:
+                x = (x + 1e-6).log()
+            x = self.instancenorm(x).unsqueeze(1)
```
Comment (XuesongYang) [inline nemo/collections/tts/modules/audio_codec_modules.py:428]: not inspect these indents thoroughly, but please double-check these lines if indents are on purposes.

File: nemo/collections/tts/modules/audio_codec_modules.py (hunk 1 above)
Comment (XuesongYang) [reply nemo/collections/tts/modules/audio_codec_modules.py:420]: still wondering why should we hack like this. `x` is `bf16`, and `with default_precision` cast it to `fp32` to call `self.torch_spec`. Outside of the block, `x` is cast back to bf16. Should we not to cast `x` to `bf16` before calling `forward`?
